content/blog/dont-use-preludes-and-globs/index.md

```diff
@@ -0,0 +1,152 @@
+
+It would be jarring to import those basic types over and over again so I'd say it's a win to have a prelude for the standard library.
+
+Maybe that's the reason why popular libraries started to provide their own preludes: [Rayon has one](https://docs.rs/rayon/latest/rayon/prelude/index.html) and so does 
+[pyo3](https://docs.rs/pyo3/latest/pyo3/prelude/index.html) and [bevy](https://docs.rs/bevy/latest/bevy/prelude/index.html) and
+[ratatui](https://docs.rs/ratatui/latest/ratatui/prelude/index.html). 
+
+Some people consider preludes and glob imports[^1] to be antipatterns &ndash; I'm one of them. Let me explain why.
+
+[^1]: A glob import (also called a wildcard import) is an import that brings all items of a module into scope. For example, `use std::collections::HashMap;` vs `use std::collections::*;`. 
+
+## I don't like preludes
+
+Preludes help beginners start quickly with zero boilerplate, but can cause namespace pollution and naming conflicts for experienced developers in larger projects.
+
+I have a hard time reviewing your code which uses preludes: looking up a definition might not work in my code review tool, and I'd have to manually track down where a types came from.
+
+Explicit imports bring clarity.
+
+Speaking of which! A good compromise between convenience and clarity is to import whole modules instead of individual items, such as `use std::fmt`. Then you can say `fmt::Debug` instead of `std::fmt::Debug`.
+
+Unless you write a highly critical framework which provides well-established types, which are *absolutely positively always* mandatory in every (or at least most) interactions with your library, don't add a prelude. Even then, make sure that prelude carries its own weight. [Tokio removed their prelude because that wasn't the case](https://github.com/tokio-rs/tokio/issues/3257).
+
+An explicit import might not be that bad. In fact, it might even clear things up for your users, because it's easier to see where imports come from.
+
+If you're worried about ergonomics, why not re-export common types in root? Then they can get imported with `use mycrate::SomeType;` instead of `use mycrate::mymodule::SomeType;`.
+
+```rust
+// Re-export a type in the your lib.rs
+pub use mymodule::SomeType;
+```
+
+On the same note, avoid glob imports (like `use prelude::*`
+or `use mycrate::mymodule::*`).
+The reasons are the same: they hide dependencies and make it harder to grep for usages of a type or function.
+
+## Common Arguments for Preludes And Glob Imports
+
+Let's address some common arguments in favor of preludes and glob imports
+and see if there's a better way to achieve the same goal.
+
+### Preludes in examples
+
+Some people like to use preludes in their documentation examples.
+
+This can be sweet when you quickly want to show how to use your library without too much boilerplate. But at what cost? At what terrible cost?[^2]
+
+[^2]: I'm exaggerating. Bear with me.
```

Comment:
The irony doesn't transport well. Also it's a very unexpected tone change, and that change doesn't add anything to the delivery or topic.

Reply:
Removed.